File: samples/sample.java
// Sample order processor — repeated identifiers exercise the rename path.
package com.example.orders;

import java.util.ArrayList;
import java.util.List;

public class OrderProcessor {

    public static class CustomerOrderLineItem {
        public String productIdentifier;
        public int quantityOrdered;
        public int unitPriceInCents;

        public CustomerOrderLineItem(String productIdentifier, int quantityOrdered, int unitPriceInCents) {
            this.productIdentifier = productIdentifier;
            this.quantityOrdered = quantityOrdered;
            this.unitPriceInCents = unitPriceInCents;
        }
    }

    public static int calculateOrderSubtotalInCents(List<CustomerOrderLineItem> customerOrderLineItems) {
        int runningSubtotalInCents = 0;
        for (CustomerOrderLineItem customerOrderLineItem : customerOrderLineItems) {
            runningSubtotalInCents += customerOrderLineItem.quantityOrdered * customerOrderLineItem.unitPriceInCents;
        }
        return runningSubtotalInCents;
    }

    public static List<CustomerOrderLineItem> filterHighValueLineItems(
            List<CustomerOrderLineItem> customerOrderLineItems, int minimumValueInCents) {
        List<CustomerOrderLineItem> highValueLineItems = new ArrayList<>();
        for (CustomerOrderLineItem customerOrderLineItem : customerOrderLineItems) {
            int lineValueInCents = customerOrderLineItem.quantityOrdered * customerOrderLineItem.unitPriceInCents;
            if (lineValueInCents >= minimumValueInCents) {
                highValueLineItems.add(customerOrderLineItem);
            }
        }
        return highValueLineItems;
    }

    public static void main(String[] args) {
        List<CustomerOrderLineItem> customerOrderLineItems = new ArrayList<>();
        customerOrderLineItems.add(new CustomerOrderLineItem("PROD-AAA", 2, 1999));
        customerOrderLineItems.add(new CustomerOrderLineItem("PROD-BBB", 1, 4999));
        customerOrderLineItems.add(new CustomerOrderLineItem("PROD-CCC", 5, 299));
        System.out.println("Subtotal: " + calculateOrderSubtotalInCents(customerOrderLineItems));
        System.out.println(
            "High value: " + filterHighValueLineItems(customerOrderLineItems, 2000).size()
        );
    }
}

File: samples/sample.js
// Sample shopping cart module — exercises rename of frequently-used identifiers.

class ShoppingCartLineItem {
  constructor(productIdentifier, quantityOrdered, unitPriceInCents) {
    this.productIdentifier = productIdentifier;
    this.quantityOrdered = quantityOrdered;
    this.unitPriceInCents = unitPriceInCents;
  }
}

function calculateShoppingCartSubtotal(shoppingCartLineItems) {
  let runningSubtotalInCents = 0;
  for (const shoppingCartLineItem of shoppingCartLineItems) {
    runningSubtotalInCents +=
      shoppingCartLineItem.quantityOrdered * shoppingCartLineItem.unitPriceInCents;
  }
  return runningSubtotalInCents;
}

function applyPercentageDiscount(shoppingCartLineItems, percentageDiscount) {
  // Mutates each line item's unitPriceInCents by the given percentage.
  for (const shoppingCartLineItem of shoppingCartLineItems) {
    const discountedPriceInCents = Math.floor(
      shoppingCartLineItem.unitPriceInCents * (1 - percentageDiscount / 100)
    );
    shoppingCartLineItem.unitPriceInCents = discountedPriceInCents;
  }
}

function summarizeShoppingCart(shoppingCartLineItems) {
  const runningSubtotalInCents = calculateShoppingCartSubtotal(shoppingCartLineItems);
  return {
    itemCount: shoppingCartLineItems.length,
    subtotalInCents: runningSubtotalInCents,
  };
}

const sampleShoppingCartLineItems = [
  new ShoppingCartLineItem("PROD-AAA", 2, 1999),
  new ShoppingCartLineItem("PROD-BBB", 1, 4999),
  new ShoppingCartLineItem("PROD-CCC", 5, 299),
];

console.log("Subtotal:", calculateShoppingCartSubtotal(sampleShoppingCartLineItems));
applyPercentageDiscount(sampleShoppingCartLineItems, 10);
console.log("After discount:", summarizeShoppingCart(sampleShoppingCartLineItems));
